server/src/test/java/org/elasticsearch/cluster/routing/allocation/AllocationFailuresResetServiceTests.java

```diff
@@ -0,0 +1,165 @@
+
+        var indexMeta = new IndexMetadata.Builder(index).settings(
+            Settings.builder().put(Settings.builder().put(IndexMetadata.SETTING_VERSION_CREATED, IndexVersion.current()).build())
+        ).numberOfShards(1).numberOfReplicas(0).build();
+
+        var meta = Metadata.builder().put(indexMeta, false).build();
+
+        var shardId = new ShardId(indexMeta.getIndex(), shard);
+        var exhaustFailures = 5;
+        var unassignedInfo = new UnassignedInfo(
+            UnassignedInfo.Reason.ALLOCATION_FAILED,
+            null,
+            null,
+            exhaustFailures,
+            0,
+            0,
+            false,
+            UnassignedInfo.AllocationStatus.NO_ATTEMPT,
+            Set.of(),
+            null
+        );
+        var shardRouting = ShardRouting.newUnassigned(
+            shardId,
+            true,
+            new RecoverySource.EmptyStoreRecoverySource(),
+            unassignedInfo,
+            ShardRouting.Role.DEFAULT
+        );
+
+        var routingTable = new RoutingTable.Builder().add(
+            new IndexRoutingTable.Builder(TestShardRoutingRoleStrategies.DEFAULT_ROLE_ONLY, indexMeta.getIndex()).initializeAsNew(
+                meta.index(index)
+            ).addIndexShard(IndexShardRoutingTable.builder(shardId).addShard(shardRouting)).build()
+        ).build();
+
+        return ClusterState.builder(ClusterName.DEFAULT).nodes(nodes).metadata(meta).routingTable(routingTable).build();
+    }
+
+    private static ClusterState addNode(ClusterState state) {
+        var nodes = DiscoveryNodes.builder(state.nodes()).add(DiscoveryNodeUtils.create("node-" + System.currentTimeMillis()));
+        return ClusterState.builder(state).nodes(nodes).build();
+    }
+
+    @Override
+    public void setUp() throws Exception {
+        super.setUp();
+        threadPool = new TestThreadPool("reset-alloc-failures");
+        clusterService = ClusterServiceUtils.createClusterService(threadPool);
+        var allocationService = new AllocationService(
+            new AllocationDeciders(List.of(new MaxRetryAllocationDecider())),
+            new TestGatewayAllocator(),
+            new BalancedShardsAllocator(Settings.EMPTY),
+            EmptyClusterInfoService.INSTANCE,
+            EmptySnapshotsInfoService.INSTANCE,
+            TestShardRoutingRoleStrategies.DEFAULT_ROLE_ONLY
+        );
+        resetService = new AllocationFailuresResetService(clusterService, allocationService);
+    }
+
+    @Override
+    public void tearDown() throws Exception {
+        super.tearDown();
+        clusterService.stop();
+        threadPool.shutdownNow();
+    }
+
+    public void testNoChangesDoesNotResetCounter() {
+        var state = clusterStateWithFailures();
+        var changeEvent = new ClusterChangedEvent("", state, state);
+        var listener = new SubscribableListener<ClusterState>();
+        resetService.processEvent(changeEvent, listener);
+        var resultState = safeAwait(listener);
```

Comment:
Consider using `DeterministicTaskQueue`:

```
var taskQueue = new DeterministicTaskQueue();
var threadPool = taskQueue.getThreadPool();
...
clusterChanged(event);
taskQueue.runAllRunnableTasks();
...
```
This way you can guarantee all tasks are executed without exposing a listener for test only purposes.

Reply:
how do I assert result? through cluster service?

Reply:
It should be accessible via `ClusterService#state`

Reply:
Couldnt make cluster service work together with deterministic queue. The reroute code path always runs after unit test code is done. After few hours of browsing other tests and try/err, I decided to use `TestThreadPool` and `assertBusy` block that will eventually get correct cluster state.